### ISSUE TITLE
Options: Unsubscribe from Contract when Dropped from Universe

### ISSUE THREAD

**Issue by jaredbroad** [QuantConnect/Lean#798]
Options/futures universes tend to have a large number of contracts. When the contracts are dropped from the universe we should unsubscribe from them to ensure IB doesn't rate limit the datafeed.

(e.g. IB typically allows subscriptions to 100,200 or 300 contracts at a time for varying subscription levels)

**Comment by buybackoff** [comment]
I am not yet fluent with Lean internals, but at the first glance, the main changes should be on [this line](https://github.com/QuantConnect/Lean/blob/master/Engine/AlgorithmManager.cs#L423). The `try` block should look like this:

```
try
{
    var changes = timeSlice.SecurityChanges;
    algorithm.OnSecuritiesChanged(changes);

    var subscriptionConfigsToRemove =
        feed.Subscriptions.Join(changes.RemovedSecurities,
            sub => sub.Security,
            sec => sec,
            (sub, sec) => sub.Configuration);

    foreach (var subscriptionConfig in subscriptionConfigsToRemove)
    {
        try
        {
            feed.RemoveSubscription(subscriptionConfig);
        }
        catch (Exception err)
        {
            Log.Error("AlgorithmManager.Run(): Cannot remove a subscription for '" + subscriptionConfig.Symbol.Value + "' in OnSecuritiesChanged event: " + err);
        }
    }
}
```

Then `feed.RemoveSubscription` method should do the job.

Is this the right direction?

**Comment by jaredbroad** [comment]
addressed by PR #943
